tempesta_fw/cache.c

```diff
@@ -248,17 +261,120 @@ tfw_cache_employ_req(TfwHttpReq *req)
+	 */
+	switch (resp->status) {
+	case 200: case 203: case 204:
+	case 300: case 301:
+	case 404: case 405: case 410: case 414:
+	case 501:
+		return true;
+	}
+	return false;
+}
+
 static bool
-tfw_cache_employ_resp(TfwHttpReq *req)
+tfw_cache_employ_resp(TfwHttpReq *req, TfwHttpResp *resp)
 {
+	static const unsigned int __read_mostly cacheit =
+		TFW_HTTP_CC_HDR_EXPIRES | TFW_HTTP_CC_MAX_AGE
+		| TFW_HTTP_CC_S_MAXAGE | TFW_HTTP_CC_PUBLIC;
+	static const unsigned int __read_mostly authcan =
+		TFW_HTTP_CC_S_MAXAGE | TFW_HTTP_CC_PUBLIC
+		| TFW_HTTP_CC_MUST_REVAL | TFW_HTTP_CC_PROXY_REVAL;
+
 	if (req->cache_ctl.flags & TFW_HTTP_CC_CFG_CACHE_BYPASS)
 		return false;
+
+	if ((req->cache_ctl.flags|resp->cache_ctl.flags) & TFW_HTTP_CC_NO_STORE)
+		return false;
+	/*
+	 * TODO: no-cache -- should be cached.
+	 * Should turn on unconditional revalidation.
+	 */
+	if (resp->cache_ctl.flags & TFW_HTTP_CC_NO_CACHE)
+		return false;
+	if (!(resp->cache_ctl.flags & TFW_HTTP_CC_IS_PRESENT)
+	    && (resp->cache_ctl.flags & TFW_HTTP_CC_PRAGMA_NO_CACHE))
+		return false;
+	if (resp->cache_ctl.flags & TFW_HTTP_CC_PRIVATE)
+		return false;
+	if ((req->cache_ctl.flags & TFW_HTTP_CC_HDR_AUTHORIZATION)
+	    && !(req->cache_ctl.flags & authcan))
+		return false;
+	if (!(resp->cache_ctl.flags & cacheit) && !tfw_cache_status_bydef(resp))
+		return false;
+
+	if (!resp->cache_ctl.timestamp)
+		resp->cache_ctl.timestamp = tfw_current_timestamp();
 	return true;
 }
 
+static time_t
+tfw_cache_calc_lifetime(TfwHttpResp *resp)
+{
+	unsigned int lifetime;
+
+	if (resp->cache_ctl.flags & TFW_HTTP_CC_S_MAXAGE)
+		lifetime = resp->cache_ctl.s_maxage;
+	else if (resp->cache_ctl.flags & TFW_HTTP_CC_MAX_AGE)
+		lifetime = resp->cache_ctl.max_age;
+	else if (resp->cache_ctl.flags & TFW_HTTP_CC_HDR_EXPIRES)
+		lifetime = resp->cache_ctl.expires;
+	else
+		lifetime = 0;	/* TODO: Heuristic lifetime. */
+
+	return lifetime;
+}
+
+static time_t
+tfw_cache_entry_age(TfwCacheEntry *ce)
+{
+	time_t apparent_age = max_t(time_t, 0, ce->resp_time - ce->date);
+	time_t corrected_age = ce->age + ce->resp_time - ce->req_time;
+	time_t initial_age = max(apparent_age, corrected_age);
+	return (initial_age + tfw_current_timestamp() - ce->resp_time);
+}
+
+static bool
+tfw_cache_entry_is_fresh(TfwHttpReq *req, TfwCacheEntry *ce)
+{
+	time_t ce_age = tfw_cache_entry_age(ce);
+	time_t ce_lifetime = ce->lifetime;
+
+	if (req->cache_ctl.flags & (TFW_HTTP_CC_MAX_AGE
+				    | TFW_HTTP_CC_MAX_STALE
+				    | TFW_HTTP_CC_MIN_FRESH))
+	{
+		time_t lt_max_age, lt_max_stale, lt_min_fresh;
+
+		lt_max_age = lt_max_stale = lt_min_fresh = UINT_MAX;
+		if (req->cache_ctl.flags & TFW_HTTP_CC_MAX_AGE)
+			lt_max_age = req->cache_ctl.max_age;
+		if (req->cache_ctl.flags & TFW_HTTP_CC_MAX_STALE)
+			lt_max_stale += req->cache_ctl.max_stale;
+		if (req->cache_ctl.flags & TFW_HTTP_CC_MIN_FRESH)
+			lt_min_fresh -= req->cache_ctl.min_fresh;
+		ce_lifetime = min(min(lt_max_age, lt_max_stale), lt_min_fresh);
+	}
+
```

Comment:
`ce_lifetime = min(ce_lifetime, ce->lifetime);` ???

And `ce_lifetime` shuld be initialized as:
```C
time_t ce_lifetime = UINT_MAX;
```